README.md
# Maubot Pocket

A [maubot](https://github.com/maubot/maubot) plugin that integrates with Pocket.

![demo](/maubot-pocket.gif)

Features:
* Authenticate with Pocket
* Fetch random article
* Archive article via reactions

If you don't want to host your own instance you can use `@maubot:basshero.org`. NOTE! This will grant 
the author of this plugin API access to your Pocket account. 

## Usage

Invite the bot (your own or `@maubot:basshero.org`) to your room.

First you need to authenticate.

`!pocket login`

Click the link, login to Pocket if needed and grant access to your account. You will be redirected
to a page that hopefully shows you have authenticated correctly.

Going back to the chat room, you can now use the `!pocket` command (without any parameters) to receive
a random article from your list.

To archive the article, react to the article with `✅` or `👎`. You can also open the article in Pocket by 
cicking the provided link, for example to add tags.

Should you want a new article you can also react to a previously posted article message with either
`➕` or `👍`.

If you want to disconnect the bot from Pocket, use the command `!pocket logout`. This will destroy
the access token from the bots database.

## Setup

### I already have a Maubot instance

Go to releases and download the latest plugin `.mbp` file. Upload it to your Maubot instance as per normal.

Create a Pocket plugin instance. Go to [Pocket developer pages](https://getpocket.com/developer/apps/) 
and create a Pocket application. Minimum required permissions are `retrieve`, `modify` - though `add` 
makes also sense since that functionality will soon be added to this bot.

Note down the "consumer key", go back to your Maubot manager and save it as the `api_key` in the Pocket plugin
instance config.

Follow above to login.

### I don't have a Maubot instance

First [install Maubot](https://docs.mau.fi/maubot/usage/setup/index.html) and then see above!

Ansible and Docker your thing? See [ansible-maubot](https://github.com/jaywink/ansible-maubot) for an Ansible role
that maintains Maubot using Docker.

## Author

Jason Robinson / https://jasonrobinson.me / `@jaywink:federator.dev`

## License

MIT


CHANGELOG.md
# CHANGELOG

## v0.2.5 (2022-04-06)

* Update maubot dependency version to v0.3.1

## v0.2.4 (2022-02-27)

* Alias 👎 to archiving an article (closes #3)

## v0.2.3 (2022-02-27)

* Add possibility to get another article by reacting with 👍 or ➕.

## v0.2.0 (2021-08-08)

* Initial version with the following features 
  * Authenticate with Pocket
  * Fetch random article
  * Archive article via reactions


pocket.py
import urllib.parse
from typing import Type, Dict

from aiohttp.web import Request, Response, json_response
from mautrix.types import RoomID
from mautrix.util.config import BaseProxyConfig, ConfigUpdateHelper
from maubot import Plugin, MessageEvent
from maubot.handlers import command, web


class Config(BaseProxyConfig):
    def do_update(self, helper: ConfigUpdateHelper) -> None:
        helper.copy("api_key")


class PocketPlugin(Plugin):
    @web.get("/authorize")
    async def authorize(self, request: Request) -> Response:
        # TODO get code and room from database
        token = "foobar"
        room_id = RoomID("!foobar:foo.bar")
        user = await self.pocket_authorize(token)
        if user.get("error_code"):
            await self.client.send_text(
                room_id,
                f"Failed to fetch user details, response code: {user.get('error_code')}",
            )
        # TODO store access token in db
        return json_response({})

    @classmethod
    def get_config_class(cls) -> Type[BaseProxyConfig]:
        return Config

    @command.new("pocket")
    async def handler(self, event: MessageEvent) -> None:
        await event.mark_read()
        # TODO raise if not authed, otherwise fetch random article
        await event.respond("Woop woop")

    @handler.subcommand(help="Authenticate with pocket")
    async def login(self, event: MessageEvent):
        # TODO raise if logged in
        token = await self.pocket_obtain_request_token()
        if token.get("error_code"):
            return await event.respond(
                f"Failed to initialize authentication flow, response code: {token.get('error_code')}",
            )
        redirect_url = urllib.parse.quote(f"{self.webapp_url}/authorize")
        await event.respond(
            f"Please continue by going to the following url and allowing access to your Pocket account: "
            f"https://getpocket.com/auth/authorize?request_token={token.get('request_token')}"
            f"&redirect_uri={redirect_url}"
        )

    @handler.subcommand(help="Disconnect from Pocket")
    async def logout(self, event: MessageEvent) -> None:
        # TODO implement
        pass

    async def pocket_authorize(self, token: str) -> Dict:
        response = await self.http.post(
            "https://getpocket.com/v3/oauth/authorize",
            headers={
                "X-Accept": "application/json",
            },
            json={
                "consumer_key": self.config["api_key"],
                "code": token,
            },
        )
        if response.status != 200:
            self.log.warning(f"Unexpected status obtaining access token: {response.status}")
            return {
                "access_token": None,
                "error_code": response.status,
                "username": None,
            }
        data = await response.json()
        if not data.get("access_token"):
            self.log.warning(f"No access token found when obtaining request token")
            return {
                "access_token": None,
                "error_code": 500,
                "username": None,
            }
        return {
            "access_token": data.get("access_token"),
            "error_code": None,
            "username": data.get("username"),
        }

    async def pocket_obtain_request_token(self) -> Dict:
        response = await self.http.post(
            "https://getpocket.com/v3/oauth/request",
            headers={
                "X-Accept": "application/json",
            },
            json={
                "consumer_key": self.config["api_key"],
                "redirect_uri": f"{self.webapp_url}/authorize",
            },
        )
        if response.status != 200:
            self.log.warning(f"Unexpected status obtaining request token: {response.status}")
            return {
                "error_code": response.status,
                "request_token": None,
            }
        data = await response.json()
        if not data.get("code"):
            self.log.warning(f"No request token found when obtaining request token")
            return {
                "error_code": 500,
                "request_token": None,
            }
        return {
            "error_code": None,
            "request_token": data.get("code"),
        }

    async def start(self) -> None:
        await super().start()
        self.config.load_and_update()


pocket/db.py
from datetime import datetime
from typing import NamedTuple, Optional

from mautrix.types import UserID, RoomID, EventID
from sqlalchemy import (
    Column, String, Integer, DateTime, Table, MetaData, select, and_,
)
from sqlalchemy.engine.base import Engine
from sqlalchemy.exc import IntegrityError

Event = NamedTuple(
    "Event",
    id=int,
    event_id=EventID,
    item_id=str,
    user_id=UserID,
)

User = NamedTuple(
    "User",
    id=int,
    user_id=UserID,
    access_token=str,
    request_room=RoomID,
    request_token=str,
    request_token_date=datetime,
    request_state=str,
)


class Database:
    db: Engine
    user: Table
    version: Table

    def __init__(self, db: Engine) -> None:
        self.db = db
        metadata = MetaData()
        self.user = Table(
            "users",
            metadata,
            Column("id", Integer, primary_key=True, autoincrement=True),
            Column("user_id", String(255), nullable=False, unique=True),
            Column("access_token", String(255), nullable=False, default=""),
            Column("request_room", String(255), nullable=False, default=""),
            Column("request_token", String(255), nullable=False, default=""),
            Column("request_token_date", DateTime, nullable=True),
            Column("request_state", String(255), nullable=False, default=""),
        )
        self.event = Table(
            "events",
            metadata,
            Column("id", Integer, primary_key=True, autoincrement=True),
            Column("event_id", String(255), nullable=False, unique=True),
            Column("item_id", String(255), nullable=False),
            Column("user_id", String(255), nullable=False),
        )
        self.version = Table(
            "version",
            metadata,
            Column("version", Integer, primary_key=True),
        )
        self.upgrade()

    def get_user_by_id(self, user_id: str) -> Optional[User]:
        rows = self.db.execute(
            select([self.user]).where(self.user.c.user_id == user_id)
        )
        try:
            row = next(rows)
            return User(*row)
        except (ValueError, StopIteration):
            return None

    def get_user_by_request_state(self, request_state: str) -> Optional[User]:
        rows = self.db.execute(
            select([self.user]).where(self.user.c.request_state == request_state)
        )
        try:
            row = next(rows)
            return User(*row)
        except (ValueError, StopIteration):
            return None

    def get_user_event(self, user_id: UserID, event_id: EventID):
        rows = self.db.execute(
            select([self.event]).where(and_(self.event.c.user_id == user_id, self.event.c.event_id == event_id))
        )
        try:
            row = next(rows)
            return Event(*row)
        except (ValueError, StopIteration):
            return None

    def set_user_access_token(self, user_id: UserID, access_token: str) -> None:
        self.db.execute(
            self.user.update()
                .where(self.user.c.user_id == user_id)
                .values(
                    access_token=access_token,
                    request_room="",
                    request_token="",
                    request_token_date=None,
                    request_state="",
                ),
        )

    def set_user_request_token(self, user_id: UserID, room_id: RoomID, request_token: str, request_state: str) -> None:
        try:
            self.db.execute(
                self.user.insert()
                    .values(
                        user_id=user_id,
                        request_room=room_id,
                        request_token=request_token,
                        request_token_date=datetime.utcnow(),
                        request_state=request_state,
                    ),
            )
        except IntegrityError:
            self.db.execute(
                self.user.update()
                    .where(self.user.c.user_id == user_id)
                    .values(
                        request_room=room_id,
                        request_token=request_token,
                        request_token_date=datetime.utcnow(),
                        request_state=request_state,
                    ),
            )

    def store_user_event(self, user_id: UserID, event_id: EventID, item_id: str) -> None:
        self.db.execute(
            self.event.insert()
                .values(
                    event_id=event_id,
                    item_id=item_id,
                    user_id=user_id,
                ),
        )

    def upgrade(self) -> None:
        self.db.execute("CREATE TABLE IF NOT EXISTS version (version INTEGER PRIMARY KEY)")
        try:
            version, = next(self.db.execute(select([self.version.c.version])))
        except (StopIteration, IndexError):
            version = 0
        if version == 0:
            self.db.execute("""CREATE TABLE IF NOT EXISTS users (
                id INTEGER NOT NULL,
                user_id VARCHAR(255) NOT NULL,
                access_token VARCHAR(255) NOT NULL DEFAULT '',
                request_room VARCHAR(255) NOT NULL DEFAULT '',
                request_token VARCHAR(255) NOT NULL DEFAULT '',
                request_token_date DATETIME NULL,
                request_state VARCHAR(255) NOT NULL DEFAULT '',
                PRIMARY KEY (id),
                UNIQUE (user_id)
            )""")
            version = 1
        if version == 1:
            self.db.execute("""CREATE TABLE IF NOT EXISTS events (
                id INTEGER NOT NULL,
                event_id VARCHAR(255) NOT NULL,
                item_id VARCHAR(255) NOT NULL,
                user_id VARCHAR(255) NOT NULL,
                PRIMARY KEY (id),
                UNIQUE (event_id)
            )""")
            version = 2
        self.db.execute(self.version.delete())
        self.db.execute(self.version.insert().values(version=version))


pocket/bot.py
import random
import urllib.parse
from typing import Type, Dict, Optional, Tuple
from uuid import uuid4

from aiohttp.web import Request, Response
from mautrix.types import UserID, EventType, ReactionEvent
from mautrix.util.config import BaseProxyConfig, ConfigUpdateHelper
from maubot import Plugin, MessageEvent
from maubot.handlers import command, web
from sqlalchemy.exc import DBAPIError

from pocket.db import Database, User


class Config(BaseProxyConfig):
    def do_update(self, helper: ConfigUpdateHelper) -> None:
        helper.copy("api_key")


class PocketPlugin(Plugin):
    db: Database

    @command.passive(
        event_type=EventType.REACTION,
        field=lambda e: e.content.relates_to.key,
        msgtypes=[],
        regex=r"\U00002795|\U0001F44D",
    )
    async def another(self, event: ReactionEvent, _: Tuple[str]) -> None:
        item_event = self.db.get_user_event(event.sender, event.content.relates_to.event_id)
        if not item_event:
            return

        user = self.db.get_user_by_id(event.sender)
        if not user or not user.access_token:
            await self.client.react(
                event.room_id,
                event.content.relates_to.event_id,
                "❌",
            )
            return

        article = await self.get_random_article(user)
        if not article:
            await self.client.react(
                event.room_id,
                event.content.relates_to.event_id,
                "🤷",
            )
            return
        event_id = await self.client.send_markdown(
            room_id=event.room_id,
            markdown=self.format_article_message(article),
        )
        self.db.store_user_event(event.sender, event_id, article["item_id"])

    @staticmethod
    def format_article_message(article):
        return f"{article['resolved_title']} - {article['resolved_url']} " \
               f"(✅/👎 to archive, [view in Pocket](https://getpocket.com/read/{article['item_id']}), 👍/➕ for another)"

    @command.passive(
        event_type=EventType.REACTION,
        field=lambda e: e.content.relates_to.key,
        msgtypes=[],
        regex=r"\U00002705|\U00002611|\U00002714|\U0001F44E",
    )
    async def archive(self, event: ReactionEvent, _: Tuple[str]) -> None:
        item_event = self.db.get_user_event(event.sender, event.content.relates_to.event_id)
        if item_event:
            user = self.db.get_user_by_id(event.sender)
            if not user or not user.access_token:
                return
            response = await self.http.post(
                "https://getpocket.com/v3/send",
                headers={
                    "X-Accept": "application/json",
                },
                json={
                    "consumer_key": self.config["api_key"],
                    "access_token": user.access_token,
                    "actions": [{
                        "action": "archive",
                        "item_id": item_event.item_id,
                    }]
                },
            )
            if response.status != 200:
                self.log.warning(f"Unexpected status archiving article: {response.status}")
                return
            self.log.info(f"Successfully archived item {item_event.item_id}")
            await self.client.react(
                event.room_id,
                event.content.relates_to.event_id,
                "🆗",
            )

    @web.get("/authorize/{request_state}")
    async def authorize(self, request: Request) -> Response:
        request_state = request.match_info.get("request_state")
        if not request_state:
            return Response(text=f"Failed to connect to Pocket, no request_state found.")

        user = self.db.get_user_by_request_state(request_state)
        if not user:
            return Response(text=f"Failed to connect to Pocket, could not match request state to a request.")

        data = await self.pocket_authorize(user.request_token)
        if data.get("error_code"):
            self.log.warning(f"Got error code from pocket authorization: {data.get('error_code')}")
            await self.client.send_notice(
                user.request_room,
                f"Failed to connect to Pocket, response code: {data.get('error_code')}",
            )
            return Response(
                text=f"Failed to connect to Pocket, response code: {data.get('error_code')}. "
                     f"Please try giving the command again.",
            )
        try:
            self.db.set_user_access_token(UserID(user.user_id), data.get("access_token"))
        except DBAPIError as ex:
            self.log.exception(f"Failed to set access token to database: {ex}")
            await self.client.send_notice(
                user.request_room,
                f"Failed to connect to Pocket due to database error. Please try again.",
            )
            return Response(
                text="Failed to connect to Pocket due to database error. Please try giving the command again.",
            )
        self.log.info(f"User {user.user_id} successfully connected to Pocket")
        await self.client.send_notice(
            user.request_room,
            f"Successfully connected to Pocket! Use `!pocket` to get a random article.",
        )
        return Response(text="Successfully connected to Pocket! You can safely close this window.")

    async def get_random_article(self, user: User) -> Optional[Dict]:
        response = await self.http.post(
            "https://getpocket.com/v3/get",
            headers={
                "X-Accept": "application/json",
            },
            json={
                "consumer_key": self.config["api_key"],
                "access_token": user.access_token,
                "detailType": "simple",
            },
        )
        if response.status != 200:
            self.log.warning(f"Unexpected status fetching random article: {response.status}")
            return
        data = await response.json()
        # noinspection PyBroadException
        try:
            articles = [item for item in data["list"].values()]
        except Exception as ex:
            self.log.warning(f"Unexpected error fetching random article: {ex}")
            return
        self.log.info(f"Got {len(articles)} articles for a user, randomizing one")
        return articles[random.randint(0, len(articles)-1)]

    @classmethod
    def get_config_class(cls) -> Type[BaseProxyConfig]:
        return Config

    @command.new("pocket", require_subcommand=False)
    async def handler(self, event: MessageEvent) -> None:
        await event.mark_read()
        user = self.db.get_user_by_id(event.sender)
        if not user or not user.access_token:
            await event.respond("You're not logged into Pocket. Try `!pocket login` first.")
            return

        article = await self.get_random_article(user)
        if not article:
            await event.respond("Didn't find any saved articles. Is your Pocket empty? Or did we hit an error?")
            return
        event_id = await event.respond(self.format_article_message(article))
        self.db.store_user_event(event.sender, event_id, article["item_id"])

    @handler.subcommand(help="Authenticate with Pocket")
    async def login(self, event: MessageEvent) -> None:
        await event.mark_read()
        user = self.db.get_user_by_id(event.sender)
        if user and user.access_token:
            await event.respond(
                "You're already logged into Pocket. Use `logout` to clear the current access token."
            )
            return
        request_state = str(uuid4())
        token = await self.pocket_obtain_request_token(request_state)
        if token.get("error_code"):
            self.log.warning(f"Failed to obtain request token, error from Pocket: {token.get('error_code')}")
            await event.respond(
                f"Failed to initialize authentication flow, response code: {token.get('error_code')}",
            )
            return
        redirect_url = urllib.parse.quote(f"{self.webapp_url}/authorize/{request_state}")
        try:
            self.db.set_user_request_token(event.sender, event.room_id, token.get("request_token"), request_state)
        except DBAPIError as ex:
            self.log.exception(f"Failed to store request token to database: {ex}")
            await event.respond("Failed to store request token, please try `login` again.")
            return
        await event.respond(
            f"Please continue by going to the following url and allowing access to your Pocket account: "
            f"https://getpocket.com/auth/authorize?request_token={token.get('request_token')}"
            f"&redirect_uri={redirect_url}"
        )

    @handler.subcommand(help="Disconnect from Pocket")
    async def logout(self, event: MessageEvent) -> None:
        await event.mark_read()
        user = self.db.get_user_by_id(event.sender)
        if not user or not user.access_token:
            await event.respond("You're not logged into Pocket.")
            return
        # TODO also nuke token on Pocket side
        try:
            self.db.set_user_access_token(event.sender, "")
        except DBAPIError as ex:
            self.log.exception(f"Failed to clear user access token: {ex}")
            await event.respond("Failed to clear access token, please contact bot admin.")
            return
        await event.respond("Successfully disconnected from Pocket.")

    async def pocket_authorize(self, token: str) -> Dict:
        response = await self.http.post(
            "https://getpocket.com/v3/oauth/authorize",
            headers={
                "X-Accept": "application/json",
            },
            json={
                "consumer_key": self.config["api_key"],
                "code": token,
            },
        )
        if response.status != 200:
            self.log.warning(f"Unexpected status obtaining access token: {response.status}")
            return {
                "access_token": None,
                "error_code": response.status,
            }
        data = await response.json()
        if not data.get("access_token"):
            self.log.warning(f"No access token found when obtaining request token")
            return {
                "access_token": None,
                "error_code": 500,
            }
        return {
            "access_token": data.get("access_token"),
            "error_code": None,
        }

    async def pocket_obtain_request_token(self, request_state: str) -> Dict:
        response = await self.http.post(
            "https://getpocket.com/v3/oauth/request",
            headers={
                "X-Accept": "application/json",
            },
            json={
                "consumer_key": self.config["api_key"],
                "redirect_uri": f"{self.webapp_url}/authorize/{request_state}",
                "state": request_state,
            },
        )
        if response.status != 200:
            self.log.warning(f"Unexpected status obtaining request token: {response.status}")
            return {
                "error_code": response.status,
                "request_token": None,
            }
        data = await response.json()
        if not data.get("code"):
            self.log.warning(f"No request token found when obtaining request token")
            return {
                "error_code": 500,
                "request_token": None,
            }
        return {
            "error_code": None,
            "request_token": data.get("code"),
        }

    async def start(self) -> None:
        await super().start()
        self.config.load_and_update()
        self.db = Database(self.database)
